### PR TITLE
Speedup connectedness by doing it on a smaller hashtable

### PR DESCRIPTION
**Pull request by michael-schwarz**
For FFmpeg, the CFG computation phase did previously not terminate in a reasonable amount of time.
With these changes (use a separate smaller hashtable to check for connectedness, this seems to work). Also this drastically reduces RAM usage, which makes me slightly suspicious

TODO:
- [ ] Check this is actually correct
- [ ] Cleanup

### CODE REVIEW

**Comment by michael-schwarz** [comment]
Most of things marked as different is due to wrapping things into calls to `Stats.time`. I'll comment on the relevant changes.

File: src/framework/cfgTools.ml [hunk 1]
```diff
@@ -156,316 +156,326 @@ let () = Printexc.register_printer (function
...
-          failwith "MyCFG.createCFG: unsupported stmt"
+  let createCFG (file: file) =
+    let cfgF = H.create 113 in
+    let cfgB = H.create 113 in
+    if Messages.tracing then Messages.trace "cfg" "Starting to build the cfg.\n\n";
+
+    let fd_nodes = NH.create 113 in
+
+    let fun_cfgF = H.create 113 in
+    let fun_cfgB = H.create 113 in
```
Comment (michael-schwarz) [inline src/framework/cfgTools.ml:167]: Hashtables that are reset after every function, used for connectedness check.

File: src/framework/cfgTools.ml [hunk 2]
```diff
@@ -156,316 +156,326 @@ let () = Printexc.register_printer (function
...
+    let addEdges fromNode edges toNode =
+      if Messages.tracing then
+        Messages.trace "cfg" "Adding edges [%a] from\n\t%a\nto\n\t%a ... "
+          pretty_edges edges
+          Node.pretty_trace fromNode
+          Node.pretty_trace toNode;
+      NH.replace fd_nodes fromNode ();
+      NH.replace fd_nodes toNode ();
+      H.add cfgB toNode (edges,fromNode); H.add fun_cfgB toNode (edges,fromNode);
+      H.add cfgF fromNode (edges,toNode); H.add fun_cfgF fromNode (edges,toNode);
```
Comment (michael-schwarz) [inline src/framework/cfgTools.ml:178]: Also add edges to per-function table

File: src/framework/cfgTools.ml [hunk 3]
```diff
@@ -156,316 +156,326 @@ let () = Printexc.register_printer (function
...
+        match glob with
+        | GFun (fd, fd_loc) ->
+          Printf.printf "Looking at the function %s.\n" fd.svar.vname;
+
+          if get_bool "dbg.cilcfgdot" then
+            Cfg.printCfgFilename ("cilcfg." ^ fd.svar.vname ^ ".dot") fd;
+
+          NH.clear fd_nodes;
+          H.clear fun_cfgB;
+          H.clear fun_cfgF;
```
Comment (michael-schwarz) [inline src/framework/cfgTools.ml:253]: Reset per function tables

**Comment by sim642** [comment]
But the connectedness check is only done using the current function's nodes using `fd_nodes` anyway, which follows the exact same clearing logic:
https://github.com/goblint/analyzer/blob/4c5135629f6f5456dcd2b57a4458c1dea985a2b0/src/framework/cfgTools.ml#L397

**Comment by michael-schwarz** [comment]
For FFmpeg, timing goes from aborting with:

```
    makeCFG                         1.047 s
    connect                        171.697 s
```

to finishing this phase with

```
    makeCFG                         7.298 s
    connect                        55.658 s
```

**Comment by sim642** [comment]
> Yes, but it still uses the large CFG hash table of all functions. We could be seeing effects where the per-function hashtable fits into a cache, and the huge one for all functions does not...

But that's far from specific to the connectedness check, Goblint has been using a single hashtable for all CFGs of all functions since the very beginning of time. And that also affects every single transfer function evaluation as well. If that's really the entirety of the bottleneck, then we have far greater problems and might have to reengineer our whole CFG representation throughout Goblint and all the downstream tools.

**Comment by sim642** [comment]
I would really love if there was a profilable reproduction of the issue (~10s runtime) that already reveals the bottleneck, because that would reveal whether the true problem is cache locality (which no part of Goblint ever considers), unsuitable hashing or something else.

**Comment by michael-schwarz** [comment]
Yes, that seems to make sense, but still doesn't quite explain the memory usage...

**Comment by sim642** [comment]
I've spent extensive amount of time profiling the counterintuitive memory usage reduction and have finally come to a conclusion.
The reason isn't something about OCaml's memory management, but rather a blatantly simple one: this changes the amount of computations (and allocations) done by `computeSCCs`! If `Cfg.next` is based on `cfgF`, then it sometimes returns more elements than `Cfg.next` based on `fun_cfgF` _when looking up the same node_! How is this possible?

Function entry (and return) nodes are the culprit: their identity (`equal` and `hash`) is based on their underlying variable (`svar`) only. So the difference appears in the form of a single function entry node having multiple successors with different `sid`s although semantically identical content. **These conflicts appear after merging of inlines was disabled in https://github.com/goblint/cil/pull/72.** Turning that on again fixes both the time and memory issue.

**Comment by sim642** [comment]
Thus, I believe that this PR alone isn't what we need. Even if we use this optimization just during the CFG construction, it does not and cannot do anything to avoid the issue in transfer functions during solving. There the same conflict would probably cause a slowdown of a similar factor (the number of unmerged copies of a function _per each unmerged copy of that function_), but to the analysis itself, not to a lightweight graph algorithm.

**Pull request closed without merging**